server/src-lib/Hasura/Events/Lib.hs

```diff
@@ -398,13 +358,67 @@ tryWebhook logenv pool e = do
       let cerr = ClientError message
       in ResponseType2 cerr
 
+    mkWebhookReq :: Value -> [HeaderConf] -> WebhookRequest
+    mkWebhookReq payload headers = WebhookRequest payload (mkMaybe headers) invocationVersion
+
+    isClientError :: Int -> Bool
+    isClientError status = status >= 1000
+
     mkMaybe :: [a] -> Maybe [a]
     mkMaybe [] = Nothing
     mkMaybe x  = Just x
 
-    isClientError :: Int -> Bool
-    isClientError status = status >= 1000
+    logQErr :: ( MonadReader r m, MonadIO m,  Has HLogger r) => QErr -> m ()
+    logQErr err = do
+      (logger:: HLogger) <- asks getter
+      liftIO $ logger $ L.toEngineLog $ EventInternalErr err
 
+    logHTTPErr :: ( MonadReader r m, MonadIO m,  Has HLogger r) => HTTPErr -> m ()
+    logHTTPErr err = do
+      (logger:: HLogger) <- asks getter
+      liftIO $ logger $ L.toEngineLog err
+
+tryWebhook
+  :: ( MonadReader r m
+     , MonadIO m
+     , MonadError HTTPErr m
+     , Has HTTP.Manager r
+     , Has HLogger r
+     , Has EventEngineCtx r
+     )
+  => [HTTP.Header] -> HTTP.ResponseTimeout -> EventPayload -> String -> m HTTPResp
+tryWebhook headers responseTimeout ep webhook = do
+  let createdAt = epCreatedAt ep
+      eventId =  epId ep
+  initReqE <- liftIO $ try $ HTTP.parseRequest webhook
+  case initReqE of
+    Left excp -> throwError $ HClient excp
+    Right initReq -> do
+      let req = initReq
+                { HTTP.method = "POST"
+                , HTTP.requestHeaders = headers
+                , HTTP.requestBody = HTTP.RequestBodyLBS (encode ep)
+                , HTTP.responseTimeout = responseTimeout
+                }
+      eeCtx <- asks getter
+      -- wait for counter and then increment beforing making http
+      liftIO $ atomically $ do
+        let EventEngineCtx _ c maxT _ = eeCtx
+        countThreads <- readTVar c
+        if countThreads >= maxT
+          then retry
+          else modifyTVar' c (+1)
+
+      eitherResp <- runHTTP req  (Just (ExtraContext createdAt eventId))
+
+      -- decrement counter once http is done
+      liftIO $ atomically $ do
+        let EventEngineCtx _ c _ _ = eeCtx
+        modifyTVar' c (\v -> v - 1)
+
+      case eitherResp of
```

Comment:
is there a one liner for this?

Reply:
Use `onLeft` from `GraphQL.Utils`. Also please move `onNothing`, `onJust`, `onLeft` from `GraphQL.Utils` to `Hasura.Prelude`